games/tictactoe/src/action_play.hh

```diff
@@ -16,7 +16,9 @@ public:
     ActionPlay() {} // for register_action()
 
     int check(const GameState& st) const override;
-    void apply(GameState& st) const override;
+    void apply(GameState* st) const override;
+    // Unhide base class apply()
+    using rules::Action<GameState>::apply;
```

Comment:
I’m not sure what you want to unhide here. The `apply` method is overiden. There is no other `apply` method. Unless I missed something.

Reply:
Ah OK I’m seen the comment on the other file. But this is a bit strange. Hopefully this is some generated code. But can't we use different names?

Reply:
Could be something like : the overridden method in protected. It is called like `apply_`. And the public `apply` method can have tons of overload if you think its usefull (shared_ptr, unique_ptr, ref, ...)

Reply:
Actually, "unhide" is not the right word.

The method is apparently not looked for by the compiler unless we explicitely surface it in the descendant class, and it has nothing to do with the overloads. This is similar to the behavior described in this article: http://www.cs.technion.ac.il/users/yechiel/c++-faq/nondependent-name-lookup-members.html

I'll reword the comments to make them clearer.

I'll check if those files can be generated.

Reply:
Did a quick test: the overload is indeed hiding the template-dependent base class method.

The solution you suggested (split overload set and implem method) is actually what is currently done, with the `apply` (overload set) and `apply_on` (implem) methods. This solution is definitvely simpler than the `using` trick. I'll revert this change.

Reply:
Updated PR, keep the apply_on() method for impl. and apply() for overloads.